### PR TITLE
FIx backward-kill-path-component behavior

### PR DESCRIPTION
**Pull request by Caroleq**
## Description

Currently when `backward-kill-path-component` is triggered with separator under the cursor, this separator is treated as a white character and consumed.  So the state machine processing components will see the first separator as a white space, next separators will be treated as normal separators.  

For example: 
| Before triggering backward-kill-path-component | After triggering backward-kill-path-component  |
| :------------------------------------------------------------:   | :-----------------------------------------------------------: |
| `~> dig example.com @`                                      | `~> dig `                                                            | 
| `~> dig example.com @@`                                   | `~> dig example.com `                                     | 
| `~> dig example.com aaaa@`                               | `~> dig example.com `                                     | 
| `~> dig example.com aaaa@@`                           | `~> dig example.com aaaa `                            | 
| `~> dig example.com aaaa@@ <spaces> `         | `~> dig example.com aaaa `                            | 
| ` ~> dig example.com  @  <spaces> `                  | `~> dig example.com  `                                    | 

PR introduces change:  When `backward-kill-path-component` is triggered with separator under the cursor, consume subsequent separators and word. 

Modified behavior of `backward-kill-path-component`:
| Before triggering backward-kill-path-component | After triggering backward-kill-path-component  |
| :------------------------------------------------------------:   | :-----------------------------------------------------------: |
| `~> dig example.com @`                                      | `~> dig example.com`                                      | 
| `~> dig example.com @@`                                   | `~> dig example.com `                                     | 
| `~> dig example.com aaaa@`                               | `~> dig example.com `                                     | 
| `~> dig example.com aaaa@@`                           | `~> dig example.com  `                                    | 
| `~> dig example.com aaaa@@ <spaces> `         | `~> dig example.com aaaa `                            | 
| ` ~> dig example.com  @  <spaces> `                  | `~> dig example.com  `                                    | 

I implemented it by adding a new state `s_initial_separator` to state machine processing components. I see no way to do it without a new state, because we need to remember if the separator we are processing is in the sequence of separators encountered on triggering `backward-kill` or was it proceeded by some non-separators.  

Fixes issue #6258

## TODOs:

- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [x] Tests have been added for regressions fixed 
- [x] User-visible changes noted in CHANGELOG.rst

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -18,6 +18,7 @@ Interactive improvements
 - When there are multiple completion candidates, fish inserts their shared prefix. This prefix was computed in a case-insensitive way, resulting in wrong case in the completion pager. This was fixed by only inserting prefixes with matching case (:issue:`7744`).
 - Commands that wrap ``cd`` (using ``complete --wraps cd``) get the same completions as ``cd`` (:issue:`4693`).
 - Arguments longer than 1024 characters no longer trigger excessive CPU usage on Mac (:issue:`7837`).
+- When ``backward-kill-path-component`` is triggered with separator under cursor, subsequent separators and a word (if there are any) will be consumed (:issue:`6258`).
```
Comment (krobelus) [inline CHANGELOG.rst:21]: Here is my attempt at rephrasing, maybe this is easier to understand?
```rest
- ``backward-kill-path-component`` (Control+W) no longer erases parts of two tokens when the cursor is positioned immediately after `` /``. (:issue:`6258`).
```

File: src/tokenizer.cpp [hunk 2]
```diff
@@ -752,18 +752,22 @@ bool move_word_state_machine_t::consume_char_path_components(wchar_t c) {
...
-    // std::fwprintf(stdout, L"state %d, consume '%lc'\n", state, c);
     bool consumed = false;
     while (state != s_end && !consumed) {
         switch (state) {
             case s_initial_punctuation: {
-                if (!is_path_component_character(c)) {
-                    consumed = true;
+                if (!is_path_component_character(c) && !iswspace(c)) {
+                    state = s_initial_separator;
+                } else{
```
Comment (krobelus) [inline src/tokenizer.cpp:765]: nit: `clang-format` will add a space here

**Review by krobelus — APPROVED** [review]
Looks great, the new behavior is definitely more logical.

File: CHANGELOG.rst (hunk 1 above)
Comment (Caroleq) [reply CHANGELOG.rst]: Agreed. I modified CHANGELOG.

File: src/tokenizer.cpp (hunk 2 above)
Comment (Caroleq) [reply src/tokenizer.cpp]: Fixed.

**Comment by krobelus** [comment]
Great work, thanks!

**Pull request merged**